Deprecate the dynamic package sensors in seventeentrack

shaiu opened the pull request:
## Breaking changes
...

## Proposed change
Add deprecation comments for the dynamic package sensors

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

shaiu committed:
Add deprecation comments for the dynamic package sensors

shaiu committed:
Add deprecation comments for the dynamic package sensors

davet2001 reviewed (dismissed):
Since the beta for 2024.5 has now been cut, the earliest deprecation start will be 2024.6.

Also is it possible to raise a deprecation warning when the deprecated functions are used?

shaiu commented:
@joostlek what did we decide on here?

joostlek commented:
I think there are 2 things we should keep in mind after I did some reviewing of the code

1. Can we get all data from the service call that was previously in the state attributes? I think we are missing the information from the individual package sensors
2. We will probably have an options field with like "create individual package sensors" that will be fixed by a repair flow. So whenever you start the integration after this is merged, a repair issue will be raised, asking you to migrate to the service calls. If you press the submit button, the options to create those sensors will turn off and then those people are migrated.

I think it would also be nice to have some example templates in the documentation to help people get started with this new way of fetching the extra data. I can ask some people if they are able to help out with this.

shaiu committed:
Add deprecation comments for the dynamic package sensors

add more information when retrieving packages from service call

shaiu committed:
Add deprecation comments for the dynamic package sensors

update deprecation comment

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

shaiu committed:
1. 17Track repair flow
2. update deprecation comment

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,78 @@
+            self.hass.config_entries.async_update_entry(self.entry, data=data)
+            return self.async_create_entry(data={})
+
+        issue_registry = ir.async_get(self.hass)
+        description_placeholders = None
+        if issue := issue_registry.async_get_issue(self.handler, self.issue_id):
+            description_placeholders = issue.translation_placeholders
```
Why do we do this?

shaiu replied:
copy paste.
removed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -113,8 +114,12 @@ async def async_setup_entry(
     coordinator: SeventeenTrackCoordinator = hass.data[DOMAIN][config_entry.entry_id]
     previous_tracking_numbers: set[str] = set()
 
+    # This has been deprecated in 2024.7, will be removed in 2025.1
```
```suggestion
    # This has been deprecated in 2024.8, will be removed in 2025.2
```

shaiu replied:
changed

shaiu committed:
1. remove description_placeholders
2. 2024.8 deprecated

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,72 @@
+
+    def __init__(self, entry: ConfigEntry) -> None:
+        """Create flow."""
+        super().__init__()
```
Is this one needed?

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,72 @@
+    ) -> data_entry_flow.FlowResult:
+        """Handle the confirm step of a fix flow."""
+        if user_input is not None:
+            data = {**self.entry.data, "deprecated": True}
```
Let's use a constant for that for now

shaiu replied:
extracted

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,72 @@
+
+    async def async_step_init(
+        self, user_input: dict[str, str] | None = None
+    ) -> data_entry_flow.FlowResult:
```
```suggestion
    ) -> FlowResult:
```

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,72 @@
+    data: dict[str, str | int | float | None] | None,
+) -> RepairsFlow:
+    """Create flow."""
+    if data and (entry_id := cast(str, data.get("entry_id"))):
+        entry = hass.config_entries.async_get_entry(entry_id)
+        assert entry
+        return SensorDeprecationRepairFlow(entry)
```
Let's also check the issue_id first

shaiu replied:
added

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -157,7 +162,9 @@ def _async_create_remove_entities():
+        _async_create_remove_entities()
 
     config_entry.async_on_unload(
         coordinator.async_add_listener(_async_create_remove_entities)
```
Should we still register this listener when its deprecated?

shaiu replied:
moved

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -45,6 +45,17 @@
+        "step": {
+          "confirm": {
+            "title": "[%key:component::seventeentrack::issues::deprecate_sensor::title%]",
+            "description": "17Track package sensors are deprecated and will be removed.\nPlease update your automations and scripts to use `Home Assistant Core Integration: Service call`."
```
Maybe we should explain what service to use

shaiu replied:
added

joostlek commented on tests/components/seventeentrack/test_repairs.py:
```diff
@@ -0,0 +1,182 @@
+    assert msg["success"]
+
+    issue = {
+        "breaks_in_ha_version": "2022.9.0dev0",
```
wait what?

shaiu replied:
changed

joostlek commented on tests/components/seventeentrack/test_repairs.py:
```diff
@@ -0,0 +1,182 @@
+
+
+@pytest.mark.usefixtures("entity_registry_enabled_by_default")
+async def test_other_fixable_issues(
```
You don't have to test this imo

shaiu replied:
I need the coverage

joostlek replied:
I rather have a missing line of code than a test that just tests to touch code

shaiu committed:
Update homeassistant/components/seventeentrack/repairs.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
1. extract deprecated to constant
2. fix types
3. check for issue_id
4. add listener only when not deprecated
5. update which service to call

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,76 @@
+from .const import DEPRECATED_KEY
+
+
+@callback
+def deprecate_sensor_issue(hass: HomeAssistant, entry_id: str) -> None:
+    """Ensure an issue is registered."""
+    ir.async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecate_sensor",
+        breaks_in_ha_version="2025.1.0",
+        issue_domain=DOMAIN,
+        is_fixable=True,
+        is_persistent=True,
+        translation_key="deprecate_sensor",
+        severity=ir.IssueSeverity.WARNING,
+        data={"entry_id": entry_id},
+    )
```
Can we move this to the place where we create it? We don't create it at many places

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,76 @@
+    ir.async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecate_sensor",
```
Let's also put the entry_id in the issue_id, so if someone set this up twice, they get 2 issues, otherwise I can see them messing up which entry they are repairng

shaiu replied:
good idea.
added

joostlek commented on homeassistant/components/seventeentrack/repairs.py:
```diff
@@ -0,0 +1,76 @@
+        hass,
+        DOMAIN,
+        "deprecate_sensor",
+        breaks_in_ha_version="2025.1.0",
```
```suggestion
        breaks_in_ha_version="2025.2.0",
```

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -45,6 +45,17 @@
+        "step": {
+          "confirm": {
+            "title": "[%key:component::seventeentrack::issues::deprecate_sensor::title%]",
+            "description": "17Track package sensors are deprecated and will be removed.\nPlease update your automations and scripts to use `Home Assistant Core Integration: Service call` `17TRACK: Get packages`."
```
```suggestion
            "description": "17Track package sensors are deprecated and will be removed.\nPlease update your automations and scripts to use the `seventeentrack.get_packages` service call."
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -157,11 +162,12 @@ def _async_create_remove_entities():
-    config_entry.async_on_unload(
-        coordinator.async_add_listener(_async_create_remove_entities)
-    )
+    if not config_entry.data.get("deprecated"):
```
use constant

shaiu replied:
done

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -113,8 +114,12 @@ async def async_setup_entry(
+    # This has been deprecated in 2024.8, will be removed in 2025.2
     @callback
     def _async_create_remove_entities():
+        if config_entry.data.get("deprecated"):
```
use cosntant

shaiu replied:
done

shaiu committed:
Update homeassistant/components/seventeentrack/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/repairs.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
1. move deprecate_sensor_issue to where needed
2. add entry_id to issue_id
3. use constant where needed

shaiu committed:
Merge branch 'dev' into shaiu-17track-deprecate-package-sensor

shaiu committed:
update breaks in ha version

joostlek commented on homeassistant/components/seventeentrack/strings.json:
```diff
@@ -45,6 +45,17 @@
+        "step": {
+          "confirm": {
+            "title": "[%key:component::seventeentrack::issues::deprecate_sensor::title%]",
+            "description": "17Track package sensors are deprecated and will be removed.\nPlease update your automations and scripts to use the `seventeentrack.get_packages` service call."
```
```suggestion
            "description": "17Track package sensors are deprecated and will be removed.\nPlease update your automations and scripts to get data using the `seventeentrack.get_packages` service call."
```

joostlek committed:
Update homeassistant/components/seventeentrack/strings.json

joostlek committed:
Remove obsolete tests

joostlek committed:
Fix

MartinHjelmare commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -157,11 +166,12 @@ def _async_create_remove_entities():
-    config_entry.async_on_unload(
-        coordinator.async_add_listener(_async_create_remove_entities)
-    )
+    if not config_entry.data.get(DEPRECATED_KEY):
```
I'd rather check for entities that are used in automations and scripts and create issues as long as those remain.

Example:
https://github.com/home-assistant/core/blob/8bc9fd23bbe9d7ddcc3d190b327cdb4a2dc4521b/homeassistant/components/habitica/sensor.py#L261-L285
